core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts

```
@@ -17,40 +17,92 @@
+      state('collapsed', style({ transform: 'rotate(0)' })),
+      transition('expanded => collapsed', animate('200ms ease-out')),
+      transition('collapsed => expanded', animate('200ms ease-in')),
+    ]),
+    trigger('languageDropdownTrigger', [
+      state('expanded', style({ transform: 'rotate(180deg)' })),
+      state('collapsed', style({ transform: 'rotate(0)' })),
+      transition('expanded => collapsed', animate('200ms ease-out')),
+      transition('collapsed => expanded', animate('200ms ease-in')),
     ]),
   ],
 })
 export class ContributorAdminDashboardPageComponent implements OnInit {
+  @ViewChild(
+    'languageDropdown', {'static': false}) languageDropdownRef!: ElementRef;
+
+  @ViewChild(
+    'activityDropdown', {'static': false}) activityDropdownRef!: ElementRef;
+
+  languageDropdownShown: boolean = false;
+  activityDropdownShown: boolean = false;
   activeTab!: string;
   TAB_NAME_TRANSLATION_SUBMITTER: string = 'Translation Submitter';
   TAB_NAME_TRANSLATION_REVIEWER: string = 'Translation Reviewer';
   TAB_NAME_QUESTION_SUBMITTER: string = 'Question Submitter';
   TAB_NAME_QUESTION_REVIEWER: string = 'Question Reviewer';
+  translationReviewersCountByLanguage!: translationReviewersCount;
   translationReviewersCount: number = 0;
   questionReviewersCount: number = 0;
-  CONTRIBUTION_TYPES!: string[];
+  CONTRIBUTION_TYPES: string[] = [];
   selectedContributionType!: string;
   isQuestionCoordinator!: boolean;
   isTranslationCoordinator!: boolean;
+  loadingMessage!: string;
+  selectedLastActivity!: number;
+  allTopicNames: string[] = [];
+
+  languageChoices: LanguageChoice[] = [];
+  lastActivity: number[] = [];
+  filter: ContributorAdminDashboardFilter = (
+    ContributorAdminDashboardFilter.createDefault());
+
+  selectedTopics: selectedTopicChoices = {
+    ids: [],
+    topics: []
+  };
+
+  selectedLanguage: LanguageChoice = {
+    language: '',
+    id: ''
+  };
```

Can we just not initialize this here? It doesn't really make sense to have a default value for a language choice like this.

---

If we don't initialize it here, it throughs an error `cannot read properties of undefined` whereever it is used.

---

Then that seems like a code smell. It doesn't really make sense to initialize a `LanguageChoice` with empty strings as the default value. Since it is possible for `selectedLanguage` to not be initialized, we should properly handle that in the code. I think initialize `selectedLanguage` to `null` and make sure the accessors of `this.selectedLanguage` handles that case properly (since the initializing `fetchAssignedLanguageIds` call may fail).

---

Implemeted the suggested flow. PTAL.

---

Well, the problem is that `null` is more clear and more accurately conveys the variable state. Initializing the fields as empty strings compiles, but aren't really valid values. Like what does it mean for an `id` or `language` to be an empty string? If you catch an error from `fetchAssignedLanguageIds` and throw there, does that satisfy the compiler?

---

Yess. The later condition you mentioned satisifes the compiler and we can throw an error there. I have already implemented this.

---

Is it possible to initialize it to `null` and check for the `null`?

---

If we initialize it to null we have to check each time we are accessing `selectedLanguage` but if we do in this way we only have to check once.

---

No, I mean adding a `catch` clause to the `fetchAssignedLanguageIds` promise to throw an error to guarantee that `selectedLanguage` is assigned a value or an error is thrown. Then, add a non-null assertion operator (`!`) to the `selectedLanguage` variable.  e.g. https://github.com/oppia/oppia/blob/develop/core/templates/components/checkpoint-celebration-modal/checkpoint-celebration-modal.component.ts#L53-L56

---

Even after doing this, we still face the behaviur I mentioned above. That is the compiler gives error of `TypeError: Cannot read properties of undefined (reading 'id')`. In case we encounter the condition you mnetioned I have added the condition to throw the error.